### ISSUE TITLE
Bug: switch-exhaustiveness-check does not consider cases with enum keys

### ISSUE THREAD

**Issue by ST-DDT** [typescript-eslint/typescript-eslint#8645]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=KYOwrgtgBAsgngUXNA3gWAFBSgQSgXigHIBDIgGk2wCEDiAjCzAX00wHp2oAlYARzABLAE7AAJlAAuAeygkANvOkB3KAGMF8gM5RlgyQAtpYSXLVrgWnTKkHgZyWAVRQkTJLgAHe-CSQAKl72hAAGACQovsjMIQDcbBgAZmAgapKC0iBSlpIAFADWwHAAXLCIyIHeAJRQ6FhQWnqSagZQBUU1ddjYGlo%2B5ZAAdDilXd3Yoo7CWQCM8fXYrAvqJH1lfhCD1KNU41CTYNNQAEzz40vjnCtrAEQkN6O2wipQIMCqCMLPwrlEAHLSUyCCCeeTACCgSTiKBwYCSUp3G7XYBEGoXbpXXr2G70B61J4vN4fL7SH7-QFQYGg8GQ6Gw%2BFQHFIrGoqDopZLdw5X5kKqxIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6KImjQA9tEjgwAXxAGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
enum MyEnum {
  A = 'a',
  B = 'b',
}

// Required to allow calls without access to the actual enum
type MyEnumType = `${MyEnum}`;

function test(key: MyEnumType) {
  switch (key) {
    case MyEnum.A: {
      return 1;
    }
    case MyEnum.B: {
      return 2;
    }
    // case "a": { throw new Error('Not implemented yet: "a" case') }
    // case "b": { throw new Error('Not implemented yet: "b" case') }
  }
}

test('a');
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/switch-exhaustiveness-check": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error is thrown, as all cases are already covered.

### Actual Result

Neither case A nor case B are taken into account.

**Comment by bradzacher** [comment]
This is working as expected right now. By doing `` `${MyEnum}` `` you've erased the enum from the type.

When we compare the two types they're no longer the same type - from our perspective you're comparing an apple and an orange.

The difference is that you're treating an enum as a translucent bag of values (I.e. You know that `MyEnum.A` resolves to `'a'`) whereas we treat an enum as an opaque one - I.e. the fact that `MyEnum.B` resolves to `'b'` is treated as a hidden implementation detail.

In general this is the more correct and safer way to operate on enums - hence there is an entire rule [`no-unsafe-enum-comparison`](https://typescript-eslint.io/rules/no-unsafe-enum-comparison/) to warn against these kinds of operations.

---

> Required to allow calls without access to the actual enum

I'm not sure I understand - why would you define an enum and then remove it from the types? Why wouldn't you export it with the function for callers to use it?

**Comment by bradzacher** [comment]
I'm not quite sure I understand - if your concern is JS users then why would it matter if you declare the type as `MyEnum` or not? As you say the type is elided so it's meaningless to them and they can in fact pass the string if they so desire.

But for TS users they would more than likely prefer the well typed enum instead because it is better for refactoring and can be a noted with JSDoc.

**Comment by ST-DDT** [comment]
> But for TS users they would more than likely prefer the well typed enum instead because it is better for refactoring and can be a noted with JSDoc.

Not sure about that. Some of my colleagues refer to the enums as added by TS as a design error.

> The difference is that you're treating an enum as a translucent bag of values (I.e. You know that MyEnum.A resolves to 'a') whereas we treat an enum as an opaque one - I.e. the fact that MyEnum.B resolves to 'b' is treated as a hidden implementation detail.

If `MyEnum.B` and `'b'` are two distinct things. Then I would expect that this rule reports them as "cannot happen".
Similar to how we already report the redundant `default` case.

AFAICT at least TS itself considers `MyEnum.B` and `'b'` to the same, because otherwise it would report it:

````ts
Type '"Foo"' is not comparable to type MyEnumType
````

**Comment by bradzacher** [comment]
> Some of my colleagues refer to the enums as added by TS as a design error.

When people say that they are referring to the unsafety of number enums and the fact that enums don't act as nominal types.

The former unsafety doesn't apply (you're using a string enum) and *most* of the latter problems have been cleaned up in newer TS versions. They're still not truly nominal types but they're very close.

---

TS does consider `'b'` and `MyEnum.B` to be of the same type in many situations. But that is piercing the opacity of the enum to its implementation detail.

I personally don't believe the rule should be piercing the opacity here.
